cluster.go

```diff
@@ -34,17 +35,29 @@ type ClusterClient struct {
 // http://redis.io/topics/cluster-spec.
 func NewClusterClient(opt *ClusterOptions) *ClusterClient {
 	client := &ClusterClient{
+		opt:     opt,
 		addrs:   opt.Addrs,
 		slots:   make([][]string, hashtag.SlotNumber),
 		clients: make(map[string]*Client),
-		opt:     opt,
 	}
 	client.commandable.process = client.process
 	client.reloadSlots()
-	go client.reaper()
 	return client
 }
 
+// getClients returns a snapshot of clients for cluster nodes
+// this ClusterClient has been working with recently.
+// Note that snapshot can contain closed clients.
+func (c *ClusterClient) getClients() map[string]*Client {
```

Comment:
I think the code gets a little messy (personal preference), how about a little helper:

```go
type clientMap struct {
	clients map[string]*Client
	mu      sync.RWMutex
}

func newClientsMap() *clientMap {
	return &clientMap{
		clients: map[string]*Client{},
	}
}

func (m *clientMap) Clients() []*Client {
	m.mu.RLock()
	clients := make([]*Client, 0, len(m.clients))
	for _, client := range m.clients {
		clients = append(clients, client)
	}
	m.mu.RUnlock()
	return clients
}

func (m *clientMap) Fetch(addr string, factory func() *Client) (*Client, error) {
	m.mu.RLock()
	client, ok := m.clients[addr]
	m.mu.RUnlock()
	if ok {
		return client, nil
	}

	m.mu.Lock()
	client, ok = m.clients[addr]
	if !ok {
		client = factory()
		m.clients[addr] = client
	}
	m.mu.Unlock()
	return client, nil
}

func (m *clientMap) Reset() (err error) {
	m.mu.Lock()
	defer m.mu.Unlock()

	for addr, client := range m.clients {
		if e := client.Close(); e != nil {
			err = e
		}
		delete(m.clients, addr)
	}
	return
}
```